torch/ao/quantization/fx/_model_report/_detector.py

```diff
@@ -134,9 +119,151 @@ def _detect_per_channel_helper(model: nn.Module):
             )
 
     if optimizations_possible:
-        further_optims_str += "To use per_channel quantization, make sure the qconfig has a per_channel weight observer."
+        further_optims_str += (
+            "To use per_channel quantization, make sure the qconfig has a per_channel weight observer."
+        )
     else:
         further_optims_str += "No further per_channel optimizations possible."
 
     # return the string and the dictionary form of same information
     return (further_optims_str, per_channel_info)
+
+
+# names for the pre and post observers that are inserted
+DEFAULT_PRE_OBSERVER_NAME = "model_report_pre_observer"
+DEFAULT_POST_OBSERVER_NAME = "model_report_post_observer"
+
+# naming conventions for stationary vs non-stationary data
+DEFAULT_STATIONARY = "stationary"
+DEFAULT_NON_STATIONARY = "non-stationary"
+
+
+def _detect_dynamic_vs_static(model: GraphModule, tolerance=0.5) -> Tuple[str, Dict[str, Any]]:
+    """
+    determines whether dynamic or static quantization is more appropriate for a given module
+
+    Stationary distribution of data are strictly above tolerance level for the comparison statistic:
+
+        S = average_batch_activation_range/epoch_activation_range
+
+    Nonstationary distributions are below the tolerance level for this metric
+
+    This will then generate suggestions for dynamic vs static quantization focused around Linear
+
+    Args:
+        model: The prepared and calibrated GraphModule with inserted ModelReportObservers around layers of interest
+
+    """
+
+    # store modules dynamic vs static information
+    module_dynamic_static_info = {}
+
+    # loop through all submodules included nested ones
+    for name, module in model.named_modules():
+        # if module has the ModelReportObserver attached to it
+        if hasattr(module, DEFAULT_PRE_OBSERVER_NAME) and hasattr(module, DEFAULT_POST_OBSERVER_NAME):
+            # get pre and post observers for the module
+            pre_obs = getattr(module, DEFAULT_PRE_OBSERVER_NAME)
+            post_obs = getattr(module, DEFAULT_POST_OBSERVER_NAME)
+
+            # get the statistics for each module
+            pre_stat = pre_obs.get_batch_to_epoch_ratio()
+            post_stat = post_obs.get_batch_to_epoch_ratio()
+
+            # record module, pre and post stat, and whether to do dynamic or static based off it
+            dynamic_recommended = False
+
+            # string for pre and post observer classification
+            pre_obs_dist_classification = ""
+            post_obs_dist_classification = ""
+
+            if pre_stat > tolerance and post_stat > tolerance:
+                dynamic_recommended = False  # static is best if both stationary
+                pre_obs_dist_classification = DEFAULT_STATIONARY
+                post_obs_dist_classification = DEFAULT_STATIONARY
```

Comment:
can we get these two just based on the relationship between pre_stat and tolerance/post_stat and tolerance? e.g.
pre_obs_dist_classification = ? if pre_stat > tolerance else ...
post_obs_dist_classification = ? if post_stat > tolerance else ...